pkg/nas/filesystem_controller.go

```diff
@@ -0,0 +1,529 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http:// www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nas
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	aliNas "github.com/aliyun/alibaba-cloud-sdk-go/services/nas"
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/kubernetes-sigs/alibaba-cloud-csi-driver/pkg/nas/cloud"
+	"github.com/kubernetes-sigs/alibaba-cloud-csi-driver/pkg/nas/internal"
+	"github.com/kubernetes-sigs/alibaba-cloud-csi-driver/pkg/utils"
+	log "github.com/sirupsen/logrus"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/record"
+)
+
+const (
+	ProtocolType    = "protocolType"
+	FileSystemType  = "fileSystemType"
+	EncryptType     = "encryptType"
+	SnapshotID      = "snapshotID"
+	StorageType     = "storageType"
+	ZoneID          = "zoneId"
+	DESCRIPTION     = "description"
+	ZoneIDTag       = "zone-id"
+	NetworkType     = "networkType"
+	VpcID           = "vpcId"
+	VSwitchID       = "vSwitchId"
+	AccessGroupName = "accessGroupName"
+	RegionID        = "regionId"
+	CnHangzhouFin   = "cn-hangzhou-finance"
+	DeleteVolume    = "deleteVolume"
+	// NASTAGKEY1 tag
+	NASTAGKEY1 = "k8s.aliyun.com"
+	// NASTAGVALUE1 value
+	NASTAGVALUE1 = "true"
+	// NASTAGKEY2 key
+	NASTAGKEY2 = "createdby"
+	// NASTAGVALUE2 value
+	NASTAGVALUE2 = "alibabacloud-csi-plugin"
+	// NASTAGKEY3 key
+	NASTAGKEY3 = "ack.aliyun.com"
+	// AddDefaultTagsError means that the add nas default tags error
+	AddDefaultTagsError string = "AddDefaultTagsError"
+	// LosetupType tag
+	LosetupType = "losetup"
+	// SkipMountType tag
+	SkipMountType = "skipmount"
+
+	csiAlibabaCloudName = "csi.alibabacloud.com"
+)
+
+var pvcMountTargetMap = map[string]string{}
```

Comment:
these three maps may has concurrency race issues?

Reply:
Changed to `sync.Map` and plan to get rid of them by using clientToken to keep idempotence in the future.